### ISSUE TITLE
[feature request] support allow-reachable-sha1-in-want

### ISSUE THREAD

**Issue by ehuss** [libgit2/libgit2#6135]
This is a feature request to support [allow-reachable-sha1-in-want](https://git-scm.com/docs/protocol-capabilities#_allow_reachable_sha1_in_want) (aka [uploadpack.allowReachableSHA1InWant](https://git-scm.com/docs/git-config#Documentation/git-config.txt-uploadpackallowReachableSHA1InWant)). If a server advertises this ability, it would be helpful to fetch a specific commit, especially if you don't know what (if any) branch it is on. This can significantly improve performance when fetching. I'm uncertain if this requires shallow support (#3058), though. 

### Reproduction steps

```c
#include "git2.h"
#include <stdio.h>
#include <assert.h>

int main() {
    assert(git_libgit2_init() == 1);

    git_repository *repo;
    assert(git_repository_init(&repo, "repo", 0) == 0);

    git_remote *remote;
    assert(git_remote_create_anonymous(&remote, repo, "https://github.com/rust-lang/cargo") == 0);

    git_fetch_options fetch_opts = GIT_FETCH_OPTIONS_INIT;

    char *strings[] = {"+88117505b8b691e0e7892630a71a85bb5e9945de:refs/success"};
    git_strarray refspecs;
    refspecs.strings = strings;
    refspecs.count = 1;

    if (git_remote_fetch(remote, &refspecs, &fetch_opts, NULL)) {
        fprintf(stderr, "failed to fetch: %s\n", git_error_last()->message);
        return 1;
    }

    // This should succeed if the commit was fetched.
    git_object *obj = NULL;
    if (git_revparse_single(&obj, repo, "refs/success")) {
        fprintf(stderr, "failed to find ref: %s\n", git_error_last()->message);
        return 1;
    }

    char oidstr[GIT_OID_HEXSZ+1] = {0};
    git_oid_tostr(oidstr, sizeof(oidstr), git_object_id(obj));
    fprintf(stderr, "oid: %s", oidstr);

    return 0;
}
```

### Expected behavior
The call to `git_remote_fetch` should fetch the specific commit.

Compared to the git CLI:

```bash
mkdir repo
cd repo
git init
git fetch --depth=1 https://github.com/rust-lang/cargo +88117505b8b691e0e7892630a71a85bb5e9945de:refs/success
git rev-parse refs/success
```

### Actual behavior
Nothing is fetched.

### Version of libgit2 (release number or SHA1)
12b53eb0318b0529514ad7e318de70b8325d43f2

### Operating system(s) tested
any

**Comment by ethomson** [comment]
Isn't this an entirely server-side limitation / feature?  I'm surprised that this doesn't work already.  🤔

**Comment by dtolnay** [comment]
> Isn't this an entirely server-side limitation / feature? I'm surprised that this doesn't work already.

Libgit2 implements fetching by filtering the list of refs offered by the server:

https://github.com/libgit2/libgit2/blob/043f3123e3c63b634d9bfd7276e70d86b8accadc/src/fetch.c#L57

If the server never offers the ref you are looking for, the filter never passes, and nothing gets fetched. Implementing "reachable sha1 in want" in libgit2 will involve sending the server a `want` followed by the SHA-1 being requested. See where it discusses "want" in https://git-scm.com/book/en/v2/Git-Internals-Transfer-Protocols.

**Comment by ethomson** [comment]
Fixed via #6203
